### PR TITLE
issue/6525-index-number-float-v2 

### PR DESCRIPTION
**Pull request by FloLey**
# Description

1. create the float type which is like a float in python. Update documentation to mention number will be deprecated

2. add testcases for following scenario and fix it

```
entity A:
    float x
end

implement A using std::none

index A(x)


std::print(A(x=0) == A(x=0.0))
```
This will still not work with the number type

3. index lookup on float with int should work

closes https://github.com/inmanta/inmanta-core/issues/6526

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: changelogs/unreleased/6526-index-float-int.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+description: number type should no longer differentiate between integer and decimal notation when used in indexes.
+issue-nr: 6526
+issue-repo: inmanta-core
+change-type: minor
+destination-branches: [master]
+sections:
+  feature: "Introduce the 'float' type which is like a float in python"
+  deprecation-note: "Deprecate the 'number' type. Instead the 'int' or 'float' type should be used"
```
Comment (sanderr) [inline changelogs/unreleased/6526-index-float-int.yml:8]: I think we should try to avoid passive voice
```suggestion
  deprecation-note: "Deprecate the 'number' type. Use the 'int' or 'float' type instead"
```

File: changelogs/unreleased/6526-index-float-int.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+description: number type should no longer differentiate between integer and decimal notation when used in indexes.
+issue-nr: 6526
+issue-repo: inmanta-core
+change-type: minor
+destination-branches: [master]
+sections:
+  feature: "Introduce the 'float' type which is like a float in python"
```
Comment (sanderr) [inline changelogs/unreleased/6526-index-float-int.yml:7]: I would not make a comparison with Python here, just say what it is.
```suggestion
  feature: "Introduce the 'float' type for floating point numbers"
```

File: docs/language.rst [hunk 3]
```diff
@@ -125,14 +125,18 @@ Literal values can be assigned to variables
 Primitive types
 ==============================
 
-The basic primitive types are ``string``, ``number``, ``int`` or ``bool``. These basic types also support type casts:
+The basic primitive types are ``string``, ``number``, ``float``, ``int`` or ``bool``. These basic types also support type casts:
+
+.. note::
+    ``number`` will be deprecated in the future
```
Comment (sanderr) [inline docs/language.rst:131]: It is already deprecated in this PR, isn't it? I would not mention deprecated types in the list above.

File: docs/language.rst [hunk 4]
```diff
@@ -125,14 +125,18 @@ Literal values can be assigned to variables
...
+
 
 .. code-block:: inmanta
 
     assert = true
     assert = int("1") == 1
-    assert = number("1.2") == 1.2
-    assert = number(true) == 1
+    assert = float("1.2") == 1.2
+    assert = float(true) == 1
```
Comment (sanderr) [inline docs/language.rst:139]:
```suggestion
    assert = int(true) == 1
```
I would not use `float` for this one:
- it makes little sense, I can't think of any use case for converting a bool to float
- the comparison to `== 1` in a casting section might make the reader believe that you would get exactly that, while you would actually get 1.0.

File: docs/language.rst [hunk 5]
```diff
@@ -164,7 +168,7 @@ For example
     typedef mac_addr as string matching /([0-9a-fA-F]{2})(:[0-9a-fA-F]{2}){5}$/
 
 
-Lists of primitive types are also primitive types: ``string[]``, ``number[]``, ``bool[]`` or ``mac_addr[]``
+Lists of primitive types are also primitive types: ``string[]``, ``float[]``, ``number[]``, ``bool[]`` or ``mac_addr[]``
```
Comment (sanderr) [inline docs/language.rst:171]:
```suggestion
Lists of primitive types are also primitive types: ``string[]``, ``float[]``, ``bool[]`` or ``mac_addr[]``
```

File: src/inmanta/ast/__init__.py [hunk 6]
```diff
@@ -382,6 +387,9 @@ def get_type(self, typ: LocatableString) -> "Type":
             else:
                 raise TypeNotFoundException(typ, self)
         elif name in self.primitives:
+            if name == "number":
+                msg = "Type 'number' is deprecated, use 'float' or 'int' instead (%s)" % str(typ.location)
+                warnings.warn(TypeDeprecationWarning(msg))
```
Comment (sanderr) [inline src/inmanta/ast/__init__.py:392]: Did you verify that you only get the warning once, no matter how many number types you use?

File: src/inmanta/ast/entity.py [hunk 7]
```diff
@@ -381,7 +381,16 @@ def add_to_index(self, instance: Instance) -> None:
...
         """
-        attributes = {k: repr(v.get_value()) for (k, v) in instance.slots.items() if v.is_ready()}
+        attributes = {}
+        for k, v in instance.slots.items():
+            if v.is_ready():
+                value = v.get_value()
+                # Check if the value is an integer,
+                # and convert to float if necessary
+                if isinstance(value, int):
+                    value = float(value)
```
Comment (sanderr) [inline src/inmanta/ast/entity.py:391]: We should not convert all ints to floats, only if the attribute type is float.

File: src/inmanta/ast/entity.py [hunk 8]
```diff
@@ -427,8 +436,10 @@ def lookup_index(
             raise NotFoundException(
                 stmt, self.get_full_name(), "No index defined on %s for this lookup: " % self.get_full_name() + str(params)
             )
-
-        key = ", ".join(["%s=%s" % (k, repr(v)) for (k, v) in sorted(params, key=lambda x: x[0])])
+        # Convert integers to floats in the key, similar to add_to_index
+        key = ", ".join(
+            ["%s=%s" % (k, repr(float(v) if isinstance(v, int) else v)) for (k, v) in sorted(params, key=lambda x: x[0])]
```
Comment (sanderr) [inline src/inmanta/ast/entity.py:441]: Same here

File: src/inmanta/ast/type.py [hunk 9]
```diff
@@ -276,11 +301,17 @@ def __init__(self) -> None:
...
-            return False
-        if not isinstance(value, numbers.Integral):
+        """
+        Validate the given value to check if it satisfies the constraints
+        associated with this type
+        """
+        if isinstance(value, AnyType):
+            return True
+
+        if not isinstance(value, int):
```
Comment (sanderr) [inline src/inmanta/ast/type.py:311]: What does the change from `numbers.Ingegral` -> `int` entail? Is this breaking?

File: src/inmanta/ast/type.py [hunk 10]
```diff
@@ -265,6 +262,34 @@ def type_string_internal(self) -> str:
...
+
+    def validate(self, value: Optional[object]) -> bool:
+        """
+        Validate the given value to check if it satisfies the constraints
+        associated with this type
+        """
+        if isinstance(value, AnyType):
+            return True
+
+        if not isinstance(value, numbers.Number):
```
Comment (sanderr) [inline src/inmanta/ast/type.py:284]: Shouldn't this be `float`?

File: src/inmanta/ast/type.py [hunk 11]
```diff
@@ -265,6 +262,34 @@ def type_string_internal(self) -> str:
...
+            raise RuntimeException(None, "Invalid value '%s', expected Float" % value)
```
Comment (sanderr) [inline src/inmanta/ast/type.py:285]:
```suggestion
            raise RuntimeException(None, "Invalid value '%s', expected float" % value)
```
Or even better: use `self.type_string()`

File: src/inmanta/ast/entity.py (hunk 7 above)
Comment (sanderr) [reply src/inmanta/ast/entity.py:391]: And use `Float.cast` to handle `Unknown`

File: changelogs/unreleased/6526-index-float-int.yml
Comment (sanderr) [inline changelogs/unreleased/6526-index-float-int.yml:1]: You should delete the old change entry.

File: src/inmanta/ast/type.py [hunk 12]
```diff
@@ -229,10 +229,7 @@ def __eq__(self, other: object) -> bool:
 @stable_api
 class Number(Primitive):
```
Comment (sanderr) [inline src/inmanta/ast/type.py:230]: You're missing the bugfix for `number()` casts and associated change entry. I understand the "Do not apply any fixes for it" part may have been misleading here. We can decide to keep the bug, but then we must also revert the fix in iso6 to keep behavior consistent. I think I prefer to just apply the fix. See [Slack](https://inmanta.slack.com/archives/CKRF0C8R3/p1700836622599319).

File: src/inmanta/ast/entity.py (hunk 7 above)
Comment (sanderr) [reply src/inmanta/ast/entity.py:391]: Actually, we don't need the index construction patch, do we? I don't really understand why you changed the index approach. The goal was to "keep `float` as in the PR".

File: src/inmanta/execute/runtime.py [hunk 13]
```diff
@@ -267,6 +267,8 @@ def set_value(self, value: T, location: Location, recur: bool = True) -> None:
                 return
         if not isinstance(value, Unknown) and self.type is not None:
             self.type.validate(value)
+        if isinstance(self.type, (Float)):
```
Comment (sanderr) [inline src/inmanta/execute/runtime.py:270]:
```suggestion
        if isinstance(self.type, Float):
```

File: tests/compiler/test_typing.py [hunk 14]
```diff
@@ -380,3 +380,241 @@ def test_2243_override_optional(snippetcompiler) -> None:
...
+        match=re.escape("Type 'number' is deprecated, use 'float' or 'int' instead"),
+    ):
+        (_, scopes) = compiler.do_compile()
+
+
+def test_number_type(snippetcompiler):
+    snippetcompiler.setup_for_snippet(
+        """
+entity Test:
+    number i = 0
```
Comment (sanderr) [inline tests/compiler/test_typing.py:404]: No need to have explicit tests for number I think. Was this meant to be a test for `float`?

File: tests/compiler/test_typing.py [hunk 15]
```diff
@@ -380,3 +380,241 @@ def test_2243_override_optional(snippetcompiler) -> None:
...
+    compiler.do_compile()
+
+
+def test_same_value_float_int(snippetcompiler):
+    snippetcompiler.setup_for_snippet(
+        """
+    i = 42.0
+    j = 42
+    i = j
+    j = i
```
Comment (sanderr) [inline tests/compiler/test_typing.py:423]: Can you also add a similar test verifying the `==` operator? i.e. `(42 == 42.0) == true`.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/ast/__init__.py (hunk 6 above)
Comment (FloLey) [reply src/inmanta/ast/__init__.py:392]: No you get it for each use of the number type. But isn't that better? I would like to see all the usages of the number type with where they are used at once so that I can make all the changes without needing to run the code multiple times.

Comment (sanderr) [reply src/inmanta/ast/__init__.py:392]: I think the solutions team will disagree very strongly. They will not be able to update immediately after this change, and they may well have hundreds of usages of this type. This would pollute their logs incredibly.

Comment (FloLey) [reply src/inmanta/ast/__init__.py:392]: But so then no need to have an indication of the line where it is used? Just one warning saying that the type is still present?

Comment (sanderr) [reply src/inmanta/ast/__init__.py:392]: Ah, I think `warnings.warn` shows each warning only once, but the message differs here for each of them because of the location information. Perhaps we can ask the solutions team how they would prefer to be warned about this?

File: changelogs/unreleased/6526-index-float-int.yml
Comment (FloLey) [reply changelogs/unreleased/6526-index-float-int.yml:1]: but the PR was reverted? it was deleted at the same time wasn't it?

File: src/inmanta/ast/__init__.py (hunk 6 above)
Comment (sanderr) [reply src/inmanta/ast/__init__.py:392]: > But so then no need to have an indication of the line where it is used? Just one warning saying that the type is still present?

I'm honestly not sure. I agree that pointing to the line is best, but I don't immediately see how to achieve both goals.

File: changelogs/unreleased/6526-index-float-int.yml
Comment (sanderr) [reply changelogs/unreleased/6526-index-float-int.yml:1]: You're absolutely right, my bad.

File: src/inmanta/ast/type.py (hunk 12 above)
Comment (FloLey) [reply src/inmanta/ast/type.py:230]: Ok i'll apply the fix. I prefer it with the fix to

File: src/inmanta/ast/type.py (hunk 9 above)
Comment (FloLey) [reply src/inmanta/ast/type.py:311]: I must say I have no idea, I didn't think of it when changing this. I'll revert to numbers.Integral, it makes me more comfortable

File: tests/compiler/test_typing.py (hunk 14 above)
Comment (FloLey) [reply tests/compiler/test_typing.py:404]: no there is already one for float, i'll remove this one.

File: src/inmanta/ast/__init__.py (hunk 6 above)
Comment (FloLey) [reply src/inmanta/ast/__init__.py:392]: It's seems even when removing the location a lot of warnings are produced. I'll need to look how to fix this

File: src/inmanta/ast/type.py (hunk 10 above)
Comment (FloLey) [reply src/inmanta/ast/type.py]: I don't think so as `isinstance(1, float) == false` and then this doesn't work anymore:

```
entity A:
    float x
end
index A(x)
a = A(x=1)
```

File: src/inmanta/ast/entity.py (hunk 7 above)
Comment (FloLey) [reply src/inmanta/ast/entity.py]: @sanderr I'm not sure I understand this comment. Do you mean that noting needs to be changed about the indexes? I thought this was the original goal?

Comment (FloLey) [reply src/inmanta/ast/entity.py]: because If I can't change the number type and I don't change the way indexes are added, I don't see how I can make this test pass:
```

def test_int_as_index_for_number(snippetcompiler):
    snippetcompiler.setup_for_snippet(
        f"""
entity A:
    number x
end
implement A using std::none
index A(x)
test = (A(x=0) == A(x=0.0))
test = true
        """,
    )
    compiler.do_compile()

```
This will always add 2 different indexes

Comment (FloLey) [reply src/inmanta/ast/entity.py]: I removed this test. It passes with floats but not with the number type

Comment (sanderr) [reply src/inmanta/ast/entity.py]: > It passes with floats but not with the number type

Perfect. We wanted to "keep float as in the PR" and "keep number as in the current stable but deprecate it. Do not apply any fixes for it, like the index matching."

File: tests/compiler/test_typing.py [hunk 16]
```diff
@@ -573,12 +558,11 @@ def test_float_type(snippetcompiler):
     assert Number().validate(u)
 
 
-@pytest.mark.parametrize_any("type", ["number", "float"])
-def test_int_as_index_for_number(snippetcompiler, type):
+def test_int_as_index_for_number(snippetcompiler):
```
Comment (sanderr) [inline tests/compiler/test_typing.py:561]:
```suggestion
def test_int_as_index_for_float(snippetcompiler):
```

File: tests/compiler/test_typing.py [hunk 17]
```diff
@@ -619,3 +603,25 @@ def test_lookup_on_int_with_float(snippetcompiler):
         """,
     )
     compiler.do_compile()
+
+
+def test_print_float(snippetcompiler, capsys):
+    snippetcompiler.setup_for_snippet(
+        """
+std::print(float(1.234))
```
Comment (sanderr) [inline tests/compiler/test_typing.py:611]: Perhaps also add a `1.0`?

**Review by sanderr — CHANGES_REQUESTED** [review]

**Comment by sanderr** [comment]
I think we also need to consider plugins, both with `float` parameters and return types.

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]

File: src/inmanta/ast/type.py [hunk 18]
```diff
@@ -590,7 +632,7 @@ class Literal(Union):
     """
 
     def __init__(self) -> None:
-        Union.__init__(self, [NullableType(Number()), Bool(), String(), TypedList(self), TypedDict(self)])
+        Union.__init__(self, [NullableType(Float()), Number(), Bool(), String(), TypedList(self), TypedDict(self)])
```
[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/ast/type.py:635]: This is the Union of all the types, By making one of them Nullable, we basically do "Float or None Or Bool or String..."
if we made them all Nullable it would be "Float or None Or Bool or None or String or None ..." So it comes down to the same. I moved it to Float as number will be gone one day.

**Comment by Hugo-Inmanta** [comment]
I've been playing around a bit with the `number` type and complex numbers:
Plugin file:

```py
from inmanta.plugins import plugin

@plugin
def cplx(re:"int", im: "int") -> "number":
    return re + im*1j
```
And model:
```
import mv2

entity A:
    number x
end
implement A using std::none

index A(x)
a = A(x=mv2::cplx(0,0))
b = A(x=mv2::cplx(2,3))

std::print(A[x=mv2::cplx(0,0)])
std::print(A[x=mv2::cplx(2,3)])
std::print(b.x)
```

Outputs:
```
@__config__::A 7f435bf269e0
@__config__::A 7f435bf26a70
@(2+3j)

```
I don't quite get why it seems to work even though direct cast to number fails
e.g.
```
c = number(mv2::cplx(2,3))

```
Fails:
```
Error: Failed to cast '(2+3j)' to number (reported in number(mv2::cplx(2,3)) (./main.cf:17))
```

File: tests/compiler/test_typing.py [hunk 19]
```diff
@@ -625,3 +572,59 @@ def test_print_number(snippetcompiler, capsys):
...
+    x = root.lookup("test").get_value()
+    assert Integer().validate(x)
+
+
+def test_float_type_argument_plugin_error(snippetcompiler, caplog):
+    snippetcompiler.setup_for_error(
+        """
+import test_674
+
+test = test_674::test_float_to_int("test")
```
Comment (sanderr) [inline tests/compiler/test_typing.py:596]: Can you parametrize this to try passing an int as well?

File: tests/data/modules/test_674/plugins/__init__.py [hunk 20]
```diff
@@ -24,3 +24,18 @@ def test_not_nullable_list(param: "int[]") -> "bool":
...
+
+
+@plugin
+def test_int_to_float(val1: "int") -> "float":
+    return float(val1)
+
+
+@plugin
+def test_error_float() -> "float":
+    return "hello"
```
Comment (sanderr) [inline tests/data/modules/test_674/plugins/__init__.py:41]: Here too I would rather opt for the `int` than `str`. No need to try both I think, just the least trivial one should suffice.

File: src/inmanta/execute/runtime.py [hunk 21]
```diff
@@ -267,6 +267,8 @@ def set_value(self, value: T, location: Location, recur: bool = True) -> None:
                 return
         if not isinstance(value, Unknown) and self.type is not None:
             self.type.validate(value)
+        if isinstance(self.type, Float):
+            value = self.type.cast(value)
```
Comment (sanderr) [inline src/inmanta/execute/runtime.py:271]: I think we should revert this change now.

**Review by sanderr — CHANGES_REQUESTED** [review]
Looks good. Final comments.

@Hugo-Inmanta I think you should find your experiments with complex numbers to be more intuitive/consistent now.

File: tests/compiler/test_typing.py [hunk 22]
```diff
@@ -643,18 +588,23 @@ def test_float_type_argument_plugin(snippetcompiler, caplog):
...
-        """
+@pytest.mark.parametrize("val", [42, "test"])
+def test_float_type_argument_plugin_error(snippetcompiler, val):
+    snippet = f"""
 import test_674
 
-test = test_674::test_float_to_int("test")
-        """,
-        ("Invalid value 'test', expected float (reported in " "test_674::test_float_to_int('test') ({dir}/main.cf:4))"),
+test = test_674::test_float_to_int({val if val == 42 else '"'+val+'"'})
```
Comment (Hugo-Inmanta) [inline tests/compiler/test_typing.py:596]: I think it reads nicer by calling `repr` which adds the single quotes for strings: e.g.

```py
    snippet = f"""
import test_674

test = test_674::test_float_to_int({repr(val)})
        """
    msg = f"Invalid value '{val}', expected float (reported in test_674::test_float_to_int({repr(val)}) ({{dir}}/main.cf:4))"
```

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]

File: tests/compiler/test_typing.py (hunk 22 above)
Comment (FloLey) [reply tests/compiler/test_typing.py:596]: Ho I didn't know that, I'll try it

**Comment by FloLey** [comment]
the compiler version are already bumped for iso6 and master

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]
One last thing, I think we should mention in the doc somewhere that our float requires a point to be valid

**Review by Hugo-Inmanta — APPROVED** [review]

**Review by sanderr — APPROVED** [review]
